packages/app/obojobo-express/obo_express_dev.js

```diff
@@ -0,0 +1,181 @@
+const sig = require('oauth-signature')
+const ltiConfig = require('./config/lti.json')
+const oauthKey = Object.keys(ltiConfig.development.keys)[0]
+const oauthSecret = ltiConfig.development.keys[oauthKey]
+
+// constructs a signed lti request and sends it.
+const renderLtiLaunch = (paramsIn, method, endpoint, res) => {
+	// add the required oauth params to the given prams
+	const oauthParams = {
+		oauth_nonce: Math.round(new Date().getTime() / 1000.0),
+		oauth_timestamp: Math.round(new Date().getTime() / 1000.0),
+		oauth_callback: 'about:blank',
+		oauth_consumer_key: oauthKey,
+		oauth_signature_method: 'HMAC-SHA1',
+		oauth_version: '1.0'
+	}
+	const params = { ...paramsIn, ...oauthParams }
+	const hmac_sha1 = sig.generate(method, endpoint, params, oauthSecret, '', {
+		encodeSignature: false
+	})
+	params['oauth_signature'] = hmac_sha1
+	const keys = Object.keys(params)
+	const htmlInput = keys
+		.map(key => `<input type="hidden" name="${key}" value="${params[key]}"/><br/>`)
+		.join('')
+
+	res.set('Content-Type', 'text/html')
+	res.send(`<html>
+		<body>
+		<form id="form" method="${method}" action="${endpoint}" >${htmlInput}</form>
+		<script>document.getElementById('form').submit()</script>
+		</body></html>`)
+}
+
+// util to get a baseUrl for inernal requests
+const baseUrl = req => `${req.protocol}://${req.get('host')}`
+
+module.exports = app => {
+	const bodyParser = require('body-parser')
+	app.use(bodyParser.json())
+	app.use(bodyParser.urlencoded({ extended: false }))
+
+	// index page with links to all the launch types
+	app.get('/lti/dev', (req, res) => {
+		res.set('Content-Type', 'text/html')
+		res.send(`<html><body>
+			<ul>
+				<li><a href="/lti/dev/launch/course_navigation">Course Nav</a></li>
+				<li><a href="/lti/dev/launch/resource_selection">Resource Selection</a></li>
+				<li><a href="/lti/dev/launch/view">View</a></li>
+				<li><a href="/profile/logout">Logout</a></li>
+			</ul>
+			</body></html>`)
+	})
+
+	// json list of every express.js route
+	app.get('/routes', (req, res) => {
+		const listEndpoints = require('express-list-endpoints')
+		const foundPaths = new Set()
+		const simplifiedEndpoints = listEndpoints(app)
+			// remove express's * path
+			.filter(i => i.path != '*')
+			// filter any duplicat paths
+			.filter(i => {
+				if(foundPaths.has(i.path)) return false
+				foundPaths.add(i.path)
+				return true;
+			})
+			// sort the remaining paths
+			.sort((a,b) => a.path.localeCompare(b.path))
+
+		res.json(simplifiedEndpoints)
+	})
+
+	// builds a valid course navigation lti launch and submits it
+	app.get('/lti/dev/launch/course_navigation', (req, res) => {
+		const method = 'POST'
+		const endpoint = `${baseUrl(req)}/lti/canvas/course_navigation`
+		const params = {
```

Comment:
would suggest making a common params object that contains user info and spreading over that inside this and subsequent route params

Reply:
Agreed